src/compound/tagged_union.ts

```diff
@@ -0,0 +1,73 @@
+import { u8 } from "../primitives/mod.ts";
+import {
+  AlignedType,
+  type InnerType,
+  type Options,
+  type Packed,
+  type ValueOf,
+} from "../types/mod.ts";
+
+type Fn<T> = (value: T) => number;
+
+export class TaggedUnion<
+  T extends Record<number, AlignedType<unknown>>,
+  V extends ValueOf<{ [K in keyof T]: InnerType<T[K]> }> = ValueOf<
+    { [K in keyof T]: InnerType<T[K]> }
+  >,
+> extends AlignedType<V> implements Packed<V> {
+  #record: T;
+  #variantFinder: Fn<V>;
+
+  constructor(input: T, variantFinder: Fn<V>) {
+    // Find biggest alignment
+    const byteAlignment = Object.values(input)
+      .reduce((acc, x) => Math.max(acc, x.byteAlignment), 0);
+    super(byteAlignment);
+    this.#record = input;
+    this.#variantFinder = variantFinder;
+  }
+
+  readUnaligned(dt: DataView, options: Options = { byteOffset: 0 }): V {
+    const discriminant = u8.readUnaligned(dt, options);
```

Comment:
```suggestion
    const discriminant = this.#discriminant.readUnaligned(dt, options);
```
The discriminant should itself be a type passed in to the constructor, not an `u8`.